src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java

```diff
@@ -202,6 +214,52 @@ public class RedundantModifierCheck
         TokenTypes.ABSTRACT,
     };
 
+    /**
+     *  Constant for jdk 21 version number.
+     */
+    private static final int JDK_21 = 21;
+
+    /**
+     *  Constant for jdk 17 version number.
+     *
+     */
+    private static final int JDK_17 = 17;
+
+    /**
+     * Set the JDK version that you are using.
+     * We support JDK versions in both notations for versions up to and including
+     * 8 (e.g., 5 and 1.5, 8 and 1.8). For versions 9 and later,
+     * use the single version number (e.g., 9, 10, 11, etc.).
+     *
+     */
+    private String jdkVersion = String.valueOf(JDK_21);
+
+    /**
+     * Integer representation of the JDK version.
+     *
+     */
+    private int jdkVersionValue = Integer.parseInt(jdkVersion);
```

Comment:
Similar to another comment, we should only need one. It can be a different type than the setter.

Reply:
I think the type in the doc is generated from the type of the field. So, we can't say `int x;` and expect it to be a String in the doc. but let me try again

Reply:
Yup, the doc and class fields have to be the same type. :(

when I changed the type to int. and leave only one field
```
     <li>
    -Property {@code jdkVersion} - Set the JDK version that you are using. We support JDK versions in both notations for versions up to and including 8 (e.g., 5 and 1.5, 8 and 1.8). For versions 9 and later, use the single version number (e.g., 9, 10, 11, etc.). Type is {@code int}. Default value is {@code 21}.
    +Property {@code jdkVersion} - Set the JDK version that you are using. We support JDK versions in both notations for versions up to and including 8 (e.g., 5 and 1.5, 8 and 1.8). For versions 9 and later, use the single version number (e.g., 9, 10, 11, etc.). Type is {@code java.lang.String}. Default value is {@code "21"}.
     </li>
```

Reply:
So thats why I have 2 differents fields one for a user-facing property (String) and the equivalent integer for us to work with.
This is a violation by PMD and IDEA inspections they said `jdkVersion` can be a local varaiable

Reply:
This also results in a pitest issue because we aren't using this variable (so removing its assignment(default value) will survive) I tried a workaround to satisfy pitest (used the `jdkVersion` field in the intialization of `jdkVersionValue` -` private int jdkVersionValue = Integer.pareInt(jdkVersion);` ), spotbugs is complaining (The object under construction remains partially initialized and may be vulnerable to Finalizer attacks.). I can look more into this to find a solution that satisfy all tools but I think supporting the single version number even when jdk <= 8 will be better at this point

Reply:
do we agree on any suppressions for any tool?

Reply:
https://github.com/checkstyle/checkstyle/pull/15376#discussion_r1706279547
> -.....Type is {@code int}.
> +.....Type is {@code java.lang.String}.

This is what we want. No suppression needed. Our type is string since we are allowing "1.X", which is not an integer.

Reply:
We can override property type display in `XDocsPagesTest`. Add property to `PROPERTIES_ALLOWED_GET_TYPES_FROM_METHOD`. Similar list is in `SiteUtil` for xdoc macro side.

Reply:
I added `RedundantModifier.jdkVersion` to both lists and still the property type is regenerated from the field (as int)
commit : https://github.com/checkstyle/checkstyle/commit/3b67a960fcd3599cbac80ad4604b61f9307a363b
CI failures: https://dev.azure.com/romanivanovjr/romanivanovjr/_build/results?buildId=23078&view=logs&j=c902ebb4-c9f8-5f09-4e17-ff78fbbc842e&t=9ca98c81-ff64-58f0-9d03-a23ac1c4a111&l=1828


why not to avoid all of these and have just a single string field and parse it to integer when needed in methods. I don't see the code getting complex with a simple `Integer.parseInt()` when needed.

Reply:
> why not to avoid all of these and have just a single string field and parse it to integer when needed in methods. I don't see the code getting complex with a simple Integer.parseInt() when needed.

I made this

Reply:
> why not to avoid all of these and have just a single string field and parse it to integer when needed in methods

I think you should try it out.

Run that version of the code on a big project. Turn halt on exception off, and set the property to "abc".

Examine what happens and also when it happens.

Reply:
We need to keep String as interface for user.
I am ok with any small performance issues, we can resolve them separately.

We should be able to introduce new class field that will keep int version, so justification of this extra field will be: metadata model defect/feature that is using class field type and not a setter type.

If extra class field is caught be some validation, we can suppress them. We just need to minimize suppressions.

May be one day we will improve metadata and this weird code will be optimized.

Reply:
Please see current code and let me now if you have any comments for now. 

I used only one field with String type.

> Run that version of the code on a big project. Turn halt on exception off, and set the property to "abc".
>
>Examine what happens and also when it happens.
>
>Then run a similar experiment with another check like JavaNCSS and set classMaximum to "abc". Both fields of these checks are internally int.

In both checks, a `NumberFormatException` is thrown

I am good with the current code we can improve it in the future if needed. adding an "extra" field that keeps the int version will cost us a lot of violations by pitest, PMD, spotbugs and IDEA inspections :).

Also making a single int field with a string setter is not working properly, metadata generation is weird and it always regenerates the type from the class field.

IMO, we can keep the current code as it is and parse the string field to int when needed. also consider that parsing to int will be in very minimal cases till now we will need it for strictfp and unnamed variables only so I think performance issues is not a big concern here.

Reply:
> In both checks, a NumberFormatException is thrown

If that is all you found, then I don't feel you did one or more of the following:
> Run that version ... on a big project
> Turn halt on exception off
> Examine if they are different and how.

=====

It has been an eternity since I have written this code. `PROPERTIES_ALLOWED_GET_TYPES_FROM_METHOD` allows us to use the setter's type ONLY if the field cannot be found. If we make the field invisible to use that override, we lose the default value of the field.

I have submitted a PR to update PROPERTIES_ALLOWED_GET_TYPES_FROM_METHOD logic so it uses the setter even if the field exists. This override is only used by 4 other checks before this PR, so it should be minimal impact.

The PR is at https://github.com/checkstyle/checkstyle/pull/15466 .

Here is your new patch. It will require the other PR to function properly and CI pass. Feel free to add better verbage.
````
diff --git a/src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java b/src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
index 804761e..5538766 100644
--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
@@ -232,7 +232,7 @@
      * use the single version number (e.g., 9, 10, 11, etc.).
      *
      */
-    private String jdkVersion = String.valueOf(JDK_21);
+    private int jdkVersion = JDK_21;
 
     /**
      * Setter to set the JDK version that you are using.
@@ -245,10 +245,10 @@
      */
     public void setJdkVersion(String jdkVersion) {
         if (jdkVersion.startsWith("1.")) {
-            this.jdkVersion = jdkVersion.substring(2);
+            this.jdkVersion = Integer.parseInt(jdkVersion.substring(2));
         }
         else {
-            this.jdkVersion = jdkVersion;
+            this.jdkVersion = Integer.parseInt(jdkVersion);
         }
     }
 
@@ -312,7 +312,7 @@
             processInterfaceOrAnnotation(ast);
         }
 
-        if (Integer.parseInt(jdkVersion) >= JDK_17) {
+        if (jdkVersion >= JDK_17) {
             checkForRedundantModifier(ast, TokenTypes.STRICTFP);
         }
     }
diff --git a/src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java b/src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java
index bbdd341..82add5d 100644
--- a/src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java
@@ -165,6 +165,8 @@
         // loads string into memory similar to file
         HEADER_CHECK_HEADER,
         REGEXP_HEADER_CHECK_HEADER,
+        // property is an int, but we cut off excess to accommodate old versions
+        "RedundantModifierCheck.jdkVersion",
         // until https://github.com/checkstyle/checkstyle/issues/13376
         "CustomImportOrderCheck.customImportOrderRules"
     );
diff --git a/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java b/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java
index 44f9ab2..813e589 100644
--- a/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java
@@ -149,6 +149,8 @@
             // loads string into memory similar to file
             "Header.header",
             "RegexpHeader.header",
+            // property is an int, but we cut off excess to accommodate old versions
+            "RedundantModifier.jdkVersion",
             // until https://github.com/checkstyle/checkstyle/issues/13376
             "CustomImportOrder.customImportOrderRules"
     );
````